### DISCUSSION TITLE
Github Actions on push not master

### DISCUSSION THREAD

**Discussion by csumpter** [community/community#26039] [Actions]
<p>I am trying to figure out what the syntax would be for running a workflow, on push, on any branch but master. I read through some other posts here that looks like folks are having issues with the tag filtering, so maybe this is related?</p>
<p>If I use code like:</p>
<pre><code>on:
  push:
    branches:
      - '*'
      - !master
</code></pre>
<p>This will throw the error:</p>
<pre><code>### ERRORED 14:48:42Z

- Your workflow file was invalid: .github/workflows/main.yml: Unexpected tag '!master'
</code></pre>
<p>Does anyone know the syntax I am looking for?</p>
<p>Thank you in advance!</p>

**Comment by csumpter** [comment]
<p>To answer my own question, the syntax appears to be:</p>
<pre><code>on:
  push:
    branches:
      - /refs/heads/*
      - !master
</code></pre>
<p>Edit: This is <strong>not</strong> the correct implementation, and only appeared to work since I had different workflow files in the master and test branches.</p>

**Comment by arpanlaha** [comment, marked as answer]
<p>I’ve found that putting quotes around !master works, but I’m not sure if it’s technically correct:</p>
<pre><code>on:
  push:
    branches:
      - '*'
      - '!master'
</code></pre>

**Reply by gerardbosch** [reply]
This answer is not correct, there's a correct one [below](https://github.com/orgs/community/discussions/26039#discussioncomment-3250144).

The `'*'` does not match branch names that contain `/`, so `fix/some-bug` won't be triggered.
`'**'` instead does the job! ✅ Thanks @stewartshea

**Comment by intolerance** [comment]
<p><a class="mention" href="/arpanlaha">@arpanlaha</a> <a class="mention" href="/csumpter">@csumpter</a> </p>
<p>We’ve been using this at the top of our <code>default.yml</code> workflow with success.</p>
<pre><code>on:
  push:
    branches:
    - !master
</code></pre>
<p>Hope that helps!</p>
<p>Edit:<br>
This apparently is no longer working – or just appeared to work somehow.</p>
<p>The good news is that <a class="mention" href="/arpanlaha">@arpanlaha</a> 's solution works fine for me.</p>

**Comment by csumpter** [comment]
<p><a class="mention" href="/arpanlaha">@arpanlaha</a> <a class="mention" href="/intolerance">@intolerance</a> </p>
<p><a class="mention" href="/arpanlaha">@arpanlaha</a>- your solution works perfectly. Thank you so much for helping out! I will flag your code as the solution.</p>

**Comment by webknjaz** [comment]
<p>Leading bangs are reserved for custom tags in YAML. So you need to always use explicit strings.</p>

**Comment by jeremyepling** [comment]
<p>You need quotes when the branch/tag starts with a ! or *. </p>

**Comment by csumpter** [comment]
<p>Thank you <a class="mention" href="/jeremyepling">@jeremyepling</a>. I see that docs have been updated on <a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions" rel="nofollow noopener">https://help.github.com/en/articles/workflow-syntax-for-github-actions</a> to reflect the same.</p>

**Comment by ghost** [comment]
<pre><code>on:
  push:
    branches:
      - '*'
      - '!master'
</code></pre>
<p>Does not seem to work for me. :frowning:</p>

**Comment by ghost** [comment]
<p>For anyone where the solution is not working. This happened to me:</p>
<p>I was pushing to a branch called ‘release/1.0.1’ and the wildcard ‘*’ does not seem to work for this naming. You also need to set ‘*/*’ to make this work.</p>
<p>As described on <a href="https://golang.org/pkg/path/#Match" rel="nofollow noopener">https://golang.org/pkg/path/#Match</a> ‘*’ : </p>
<pre><code>matches any sequence of non-/ characters 
</code></pre>

**Comment by stewartshea** [comment]
<p>Similarly to the problem that <a class="mention" href="/verbruggenalex">@verbruggenalex</a> highlighted, the double ** will work as well:</p>
<pre><code class="lang-auto">on:
  push:
    branches:
      - '**'
      - '!master' 
    paths: 
    - 'somepath/**'
</code></pre>
<p>The documentation <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet" rel="nofollow noopener">here</a> states:</p>
<ul>
<li>
<code>*</code> : Matches zero or more characters, but does not match the  <code>/</code>  character. For example,  <code>Octo*</code>  matches  <code>Octocat</code> .</li>
<li>
<code>**</code> : Matches zero or more of any character.</li>
</ul>

**Comment by arpanlaha** [comment]
<p>Thanks for this, hopefully this can be marked as the solution instead of what I had previously.</p>

**Comment by cascandaliato** [comment]
<p>Using <code>branches-ignore</code> from the <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-ignoring-branches-and-tags" rel="nofollow noopener">official docs</a>:</p>
<pre><code class="lang-yaml">on:
  push:
    branches-ignore:
      - master
</code></pre>

**Reply by cowwoc** [reply]
"This is the way"
